File: app/src/main/java/com/stedi/randomimagegenerator/app/view/dialogs/ConfirmDialog.java
package com.stedi.randomimagegenerator.app.view.dialogs;

import android.app.Dialog;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AlertDialog;

import com.stedi.randomimagegenerator.app.R;
import com.stedi.randomimagegenerator.app.other.CachedBus;
import com.stedi.randomimagegenerator.app.other.ExtKt;
import com.stedi.randomimagegenerator.app.view.dialogs.base.BaseDialogFragment;

import javax.inject.Inject;

public class ConfirmDialog extends BaseDialogFragment {
    private static final String KEY_REQUEST_CODE = "KEY_REQUEST_CODE";
    private static final String KEY_TITLE = "KEY_TITLE";
    private static final String KEY_MESSAGE = "KEY_MESSAGE";

    @Inject CachedBus bus;

    private int requestCode;
    private boolean posted;

    public static class Callback {
        public final int requestCode;
        public final boolean confirm;

        Callback(int requestCode, boolean confirm) {
            this.requestCode = requestCode;
            this.confirm = confirm;
        }
    }

    @NonNull
    public static ConfirmDialog newInstance(int requestCode, @Nullable String title, @Nullable String message) {
        Bundle args = new Bundle();
        args.putInt(KEY_REQUEST_CODE, requestCode);
        args.putString(KEY_TITLE, title);
        args.putString(KEY_MESSAGE, message);
        ConfirmDialog dlg = new ConfirmDialog();
        dlg.setArguments(args);
        return dlg;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ExtKt.getApp(getContext()).getComponent().inject(this);
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        requestCode = getArguments().getInt(KEY_REQUEST_CODE);
        AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
        String title = getArguments().getString(KEY_TITLE, null);
        if (title != null)
            builder.setTitle(title);
        String message = getArguments().getString(KEY_MESSAGE, null);
        if (message != null)
            builder.setMessage(message);
        builder.setPositiveButton(R.string.ok, (dialog, which) -> {
            posted = true;
            bus.postDead(new Callback(requestCode, true));
        });
        builder.setNegativeButton(R.string.cancel, null);
        return builder.create();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (!posted)
            bus.postDead(new Callback(requestCode, false));
    }
}

File: app/src/main/java/com/stedi/randomimagegenerator/app/view/dialogs/ConfirmDialog.kt
package com.stedi.randomimagegenerator.app.view.dialogs

import android.app.Dialog
import android.os.Bundle
import android.support.v7.app.AlertDialog
import com.stedi.randomimagegenerator.app.R
import com.stedi.randomimagegenerator.app.other.CachedBus
import com.stedi.randomimagegenerator.app.view.activity.base.BaseActivity
import com.stedi.randomimagegenerator.app.view.dialogs.base.BaseDialogFragment
import javax.inject.Inject

class ConfirmDialog : BaseDialogFragment() {

    @Inject lateinit var bus: CachedBus

    private var requestCode: Int = 0
    private var posted: Boolean = false

    class Callback(val requestCode: Int, val confirm: Boolean)

    companion object {
        private const val KEY_REQUEST_CODE = "KEY_REQUEST_CODE"
        private const val KEY_TITLE = "KEY_TITLE"
        private const val KEY_MESSAGE = "KEY_MESSAGE"

        fun newInstance(requestCode: Int, title: String?, message: String?): ConfirmDialog {
            return ConfirmDialog().apply {
                arguments = Bundle().apply {
                    putInt(KEY_REQUEST_CODE, requestCode)
                    putString(KEY_TITLE, title)
                    putString(KEY_MESSAGE, message)
                }
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (activity as BaseActivity).activityComponent.inject(this)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        return AlertDialog.Builder(context!!).apply {
            requestCode = arguments?.getInt(KEY_REQUEST_CODE) ?: 0

            val title = arguments?.getString(KEY_TITLE, null)
            if (title != null) {
                setTitle(title)
            }

            val message = arguments?.getString(KEY_MESSAGE, null)
            if (message != null) {
                setMessage(message)
            }

            setPositiveButton(R.string.ok) { _, _ ->
                posted = true
                bus.postDeadEvent(Callback(requestCode, true))
            }
            setNegativeButton(R.string.cancel, null)
        }.create()
    }

    override fun onDestroy() {
        super.onDestroy()
        if (!posted) {
            bus.postDeadEvent(Callback(requestCode, false))
        }
    }
}